File: src/main/java/me/taubsie/carrylogs/application/service/MessagesService.java
package me.taubsie.carrylogs.application.service;

import me.taubsie.carrylogs.application.connection.DungeonHubConnection;
import me.taubsie.carrylogs.application.enums.EmbedColor;
import me.taubsie.carrylogs.application.start.BotStarter;
import me.taubsie.dungeonhub.common.*;
import org.javacord.api.entity.channel.ServerTextChannel;
import org.javacord.api.entity.message.Message;
import org.javacord.api.entity.message.embed.EmbedBuilder;
import org.javacord.api.entity.server.Server;

import java.sql.Time;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@OnStart
public class MessagesService implements StartupListener {
    private static final long REFRESH_PERIOD = 1000L * 60 * 15;
    private static MessagesService instance;

    public static MessagesService getInstance() {
        if (instance == null) {
            instance = new MessagesService();
        }

        return instance;
    }

    public Optional<EmbedBuilder> getPriceEmbed(CarryTier carryTier) {
        List<CarryDifficulty> carryDifficulties = DungeonHubConnection.getInstance().loadCarryDifficulties(carryTier);

        return getEmbedFromCarryDifficulty(carryDifficulties);
    }

    private Optional<EmbedBuilder> getEmbedFromCarryDifficulty(List<CarryDifficulty> carryDifficulties) {
        if (carryDifficulties.isEmpty()) {
            return Optional.empty();
        }

        CarryDifficulty mainCarryDifficulty = carryDifficulties.get(0);

        String title = "## " + mainCarryDifficulty.getCarryTier().getPriceTitle() + "\n";

        String description = title + carryDifficulties.stream()
                .map(carryDifficulty -> {
                    StringBuilder result = new StringBuilder();

                    if (carryDifficulty.getBulkAmount().isPresent() && carryDifficulty.getBulkPrice().isPresent()) {
                        result.append("\n");
                    }

                    result.append("**")
                            .append(carryDifficulty.getPriceName())
                            .append("**: ")
                            .append(ApplicationService.getInstance().makeNumberReadable(carryDifficulty.getPrice()))
                            .append(" coins");

                    if (carryDifficulty.getBulkAmount().isPresent() && carryDifficulty.getBulkPrice().isPresent()) {
                        result.append("\n\\*")
                                .append(ApplicationService.getInstance().makeNumberReadable(carryDifficulty.getBulkPrice().get()))
                                .append(" per carry if you buy ")
                                .append(carryDifficulty.getBulkAmount().get())
                                .append("+ carries.");
                    }

                    return result;
                }).collect(Collectors.joining("\n"));

        EmbedBuilder embed = ApplicationService.getInstance()
                .getEmbedWithoutTimestamp()
                .setColor(EmbedColor.DEFAULT.getColor())
                .setDescription(description);

        if (mainCarryDifficulty.getCarryTier().getThumbnailUrl().isPresent()) {
            embed.setThumbnail(mainCarryDifficulty.getCarryTier().getThumbnailUrl().get());
        }

        return Optional.of(embed);
    }

    public void refreshPriceMessages(Server server) {
        refreshPriceMessages(DungeonHubConnection.getInstance().loadCarryTiers(server).stream());
    }

    private void refreshPriceMessages() {
        refreshPriceMessages(DungeonHubConnection.getInstance().loadCarryTiers().stream());
    }

    private void refreshPriceMessages(Stream<CarryTier> carryTiers) {
        Map<Long, List<CarryTier>> carryTiersPerChannel = carryTiers
                .filter(carryTier -> carryTier.getPriceChannel().isPresent())
                .collect(Collectors.toMap(
                        carryTier -> carryTier.getPriceChannel().get(),
                        carryTier -> new ArrayList<>(List.of(carryTier)),
                        (o, o2) -> {
                            o.addAll(o2);
                            return o;
                        }
                ));

        carryTiersPerChannel
                .forEach((key, value) -> BotStarter.getInstance().getBot()
                        .getServerTextChannelById(key)
                        .ifPresent(serverTextChannel -> refreshPriceMessageInChannel(
                                serverTextChannel,
                                addPriceFooterToLast(
                                        value.stream()
                                                .flatMap(carryTier -> getPriceEmbed(carryTier).stream())
                                                .toList()
                                )
                        ))
                );
    }

    private List<EmbedBuilder> addPriceFooterToLast(List<EmbedBuilder> embeds) {
        for(EmbedBuilder embed : embeds) {
            embed.setFooter(null);
        }

        if(embeds.size() > 0) {
            embeds.get(embeds.size() - 1).setFooter(ApplicationService.getInstance().getPriceFooter());
        }

        return embeds;
    }

    private void refreshPriceMessageInChannel(ServerTextChannel textChannel, List<EmbedBuilder> embeds) {
        if (embeds.isEmpty()) {
            return;
        }

        Optional<Message> messageOptional =
                textChannel.getMessagesAsStream().filter(message -> message.getAuthor().isYourself()).findFirst();

        if (messageOptional.isEmpty()) {
            textChannel.sendMessage(embeds);
        } else {
            messageOptional.get().createUpdater().removeAllEmbeds().addEmbeds(embeds).applyChanges().join();
        }
    }

    @Override
    public void postStart(ProgramOrigin programOrigin) {
        new Timer().scheduleAtFixedRate(new TimerTask() {
            @Override
            public void run() {
                refreshPriceMessages();
            }
        }, new Time(System.currentTimeMillis() + 15000), REFRESH_PERIOD);
    }
}

File: src/main/kotlin/net/dungeonhub/application/service/MessagesService.kt
package net.dungeonhub.application.service

import dev.kord.core.behavior.channel.createMessage
import dev.kord.core.behavior.edit
import dev.kord.core.entity.channel.GuildMessageChannel
import dev.kord.rest.builder.message.EmbedBuilder
import kotlinx.coroutines.flow.firstOrNull
import net.dungeonhub.application.enums.EmbedColor
import net.dungeonhub.connection.CarryDifficultyConnection
import net.dungeonhub.model.carry_difficulty.CarryDifficultyModel
import net.dungeonhub.model.carry_tier.CarryTierModel
import java.util.stream.Collectors

// TODO either: delete this, or add some other embeds from e.g. StaticMessageService into here
object MessagesService {
    fun getPriceEmbed(carryTier: CarryTierModel): EmbedBuilder? {
        val carryDifficulties = CarryDifficultyConnection[carryTier].authenticated().allCarryDifficulties ?: listOf()

        if (carryDifficulties.isEmpty()) {
            return null
        }

        val title = "## " + carryTier.priceTitle + "\n"
        val priceDescription = carryTier.priceDescription

        val description =
            title + (priceDescription?.let { s: String -> s + "\n\n" } ?: "") + carryDifficulties.stream()
                .map { carryDifficulty: CarryDifficultyModel ->
                    val result = StringBuilder()
                    if (carryDifficulty.bulkAmount != null && carryDifficulty.bulkPrice != null) {
                        result.append("\n")
                    }

                    result.append("**")
                        .append(carryDifficulty.priceName)
                        .append("**: ")

                    val priceText = if (carryDifficulty.price != 0
                    ) ApplicationService.makeNumberReadable(carryDifficulty.price.toLong()) + " coins"
                    else "Free"

                    result.append(priceText)

                    if (carryDifficulty.bulkAmount != null && carryDifficulty.bulkPrice != null) {
                        result.append("\n\\*")
                            .append(ApplicationService.makeNumberReadable(carryDifficulty.bulkPrice!!.toLong()))
                            .append(" per carry if you buy ")
                            .append(carryDifficulty.bulkAmount)
                            .append("+ carries.")
                    }
                    result
                }.collect(Collectors.joining("\n"))

        val embed = ApplicationService.embedWithoutTimestamp
        embed.color = EmbedColor.Default.color
        embed.description = description

        carryTier.thumbnailUrl?.let { embed.thumbnail { this.url = it } }

        return embed
    }

    private fun addPriceFooterToLast(embeds: List<EmbedBuilder>): List<EmbedBuilder> {
        for (embed in embeds) {
            embed.footer { text = "" }
        }

        if (embeds.isNotEmpty()) {
            embeds[embeds.size - 1].footer { text = ApplicationService.priceFooter }
        }

        return embeds
    }

    private suspend fun refreshPriceMessageInChannel(
        textChannel: GuildMessageChannel,
        embeds: List<EmbedBuilder>
    ) {
        if (embeds.isEmpty()) {
            return
        }

        val message = textChannel.messages.firstOrNull { it.author?.isSelf == true }

        if (message == null) {
            textChannel.createMessage { this@createMessage.embeds = embeds.toMutableList() }
        } else {
            message.edit { this@edit.embeds = embeds.toMutableList() }
        }
    }
}